### PR TITLE
Bitcoin transactions frontend

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/32709

- Removed some code which appeared to be not needed for bitcoin support.
- Backend support of AddUnapprovedTransaction/ApproveTransaction calls.
- Frontend support of displaying BTC transaction

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/bitcoin/bitcoin_wallet_service_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+    brave_wallet::RegisterLocalStatePrefs(local_state_.registry());
+
+    keyring_service_ =
+        std::make_unique<KeyringService>(nullptr, &prefs_, &local_state_);
+    bitcoin_wallet_service_ = std::make_unique<BitcoinWalletService>(
+        keyring_service_.get(), &prefs_, shared_url_loader_factory_);
+
+    url_loader_factory_.SetInterceptor(
+        base::BindRepeating(&BitcoinWalletServiceUnitTest::RequestInterceptor,
+                            base::Unretained(this)));
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/bitcoin/bitcoin_wallet_service_unittest.cc:75]: `url_loader_factory_` is owned by `this`

File: components/brave_wallet/browser/bitcoin/bitcoin_wallet_service.h [hunk 2]
```diff
@@ -42,59 +43,88 @@ class BitcoinWalletService : public KeyedService,
...
+                GetUtxosCallback callback);
+
+  using CreateTransactionCallback =
+      base::OnceCallback<void(base::expected<BitcoinTransaction, std::string>)>;
+  void CreateTransaction(const std::string& chain_id,
+                         mojom::AccountIdPtr account_id,
+                         const std::string& address_to,
+                         uint64_t amount,
+                         CreateTransactionCallback callback);
+  void CreateTransactionTaskDone(CreateTransactionTask* task);
```
Comment (darkdh) [inline components/brave_wallet/browser/bitcoin/bitcoin_wallet_service.h:78]: can we make it private and friend `CreateTransactionTask`?

File: components/brave_wallet/browser/bitcoin/bitcoin_wallet_service_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+TEST_F(BitcoinWalletServiceUnitTest, GetBalance) {
+  base::MockCallback<BitcoinWalletService::GetBalanceCallback> callback;
+
+  // GetBalance works.
+  auto expected_balance = mojom::BitcoinBalance::New();
+  expected_balance->total_balance = 128332;
+  for (auto& addr : address_stats_map_) {
+    expected_balance->balances[addr.first] = 0;
+  }
+  expected_balance->balances[address_0_] = 11666;
```
Comment (darkdh) [inline components/brave_wallet/browser/bitcoin/bitcoin_wallet_service_unittest.cc:234]: can we write this as 10000 - 5000 + 8888 - 2222?

**Review by darkdh — APPROVED** [review]
great work!

File: components/brave_wallet_ui/common/hooks/send.ts [hunk 4]
```diff
@@ -332,6 +332,25 @@ export function useSend (isSendTab?: boolean) {
...
+      // eslint-disable-next-line no-useless-return
+      return
+    }
+
+    // Check if value is the same as the sending address
+    // TODO(apaymyshev): should prohibit self transfers?
+
+    // TODO(apaymyshev): validate address format.
+  }, [])
+
```
Comment (onyb) [inline components/brave_wallet_ui/common/hooks/send.ts:353]: nit: we're moving away from unnecessary memos and callbacks, which don't have the kind of performance benefits as previously thought.

File: components/brave_wallet_ui/components/extension/transaction-box/bitcoin-transaction-detail-box.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// types
+import { BraveWallet } from '../../../constants/types'
+
+import {
+  CodeSnippet,
+  CodeSnippetText,
+  DetailColumn,
+} from './style'
+
+export interface Props {
```
Comment (onyb) [inline components/brave_wallet_ui/components/extension/transaction-box/bitcoin-transaction-detail-box.tsx:20]: `export` not necessary

File: components/brave_wallet/common/brave_wallet.mojom [hunk 6]
```diff
@@ -1365,8 +1372,13 @@ struct BitcoinAccountInfo {
...
 };
 
+struct BitcoinBalance {
+  uint64 total_balance;
+  map<string, uint64> balances;
+};
+
 interface BitcoinWalletService {
-  GetBalance(string network_id, AccountId account_id) => (string balance, string error_message);
+  GetBalance(string network_id, AccountId account_id) => (BitcoinBalance? balance, string? error_message);
```
Comment (onyb) [inline components/brave_wallet/common/brave_wallet.mojom:1381]: Let's add some comments for `GetBalance`. For example, are unconfirmed balances included?

File: components/brave_wallet_ui/common/slices/api.slice.ts [hunk 7]
```diff
@@ -1716,11 +1717,60 @@ export function createWalletApi () {
...
+        SendBtcTransactionParams
+      >({
+        queryFn: async (payload, { dispatch }, extraOptions, baseQuery) => {
+          try {
+            const { txService } = baseQuery(undefined).data
+
+            const btcTxData: BraveWallet.BtcTxData = {
+              to: payload.to,
+              amount: BigInt(payload.value),
+              fee: BigInt(0),
```
Comment (onyb) [inline components/brave_wallet_ui/common/slices/api.slice.ts:1731]: Why is this field `0`? Also, shouldn't this field be called `feePerByte`?

File: components/brave_wallet_ui/common/slices/api.slice.ts [hunk 8]
```diff
@@ -1716,11 +1717,60 @@ export function createWalletApi () {
...
+              inputs: [],
+              outputs: []
```
Comment (onyb) [inline components/brave_wallet_ui/common/slices/api.slice.ts:1733]: I'm assuming these are automatically populated by the backend?

**Review by onyb — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/common/hooks/send.ts (hunk 4 above)
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/send.ts:353]: Interesting talk on this topic: https://youtu.be/ga4qoNJGNts?si=IY79TeEOBs484_wA

Comment (supermassive) [reply components/brave_wallet_ui/common/hooks/send.ts:353]: Just being consistent with processing addresses for other coins.

And thanks for the link!

File: components/brave_wallet_ui/common/slices/api.slice.ts (hunk 7 above)
Comment (supermassive) [reply components/brave_wallet_ui/common/slices/api.slice.ts:1731]: Not used by backend.

**Review by onyb — APPROVED** [review]
LGTM ✅

**Pull request merged**